Riot/Modules/Home/AllChats/AllChatsViewController.swift

```diff
@@ -87,41 +104,230 @@ class AllChatsViewController: HomeViewController {
+    
+    private var lastScrollPosition: Double = 0
+
+    override func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
+        lastScrollPosition = self.recentsTableView.contentOffset.y
+    }
+
+    override func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        super.scrollViewDidScroll(scrollView)
+        
+        if self.recentsTableView.contentOffset.y == 0 {
+            self.navigationController?.setToolbarHidden(false, animated: true)
+        }
+
+        guard self.recentsTableView.isDragging else {
+            return
+        }
+
+        let scrollPosition = max(self.recentsTableView.contentOffset.y, 0)
+        guard scrollPosition < self.recentsTableView.contentSize.height - self.recentsTableView.bounds.height else {
+            return
+        }
+
+        self.navigationController?.setToolbarHidden(scrollPosition - lastScrollPosition > 0, animated: true)
+        lastScrollPosition = scrollPosition
+    }
+    
     // MARK: - Private
     
     @objc private func setupEditOptions() {
-        // Note: updating toolbar items doesn't work as expected and has weird behaviour
-        // Also this piece of code is going to be updated in the next PR
-        
-        let editMenu = UIMenu(children: [
-            UIAction(title: VectorL10n.roomRecentsJoinRoom,
-                     image: Asset.Images.homeFabJoinRoom.image,
-                     discoverabilityTitle: VectorL10n.roomRecentsJoinRoom,
-                     handler: { [weak self] action in
-                self?.joinARoom()
-            }),
-            UIAction(title: VectorL10n.roomRecentsCreateEmptyRoom,
-                     image: Asset.Images.homeFabCreateRoom.image,
-                     discoverabilityTitle: VectorL10n.roomRecentsCreateEmptyRoom,
-                     handler: { [weak self] action in
-                self?.createNewRoom()
-            }),
-            UIAction(title: VectorL10n.roomRecentsStartChatWith,
-                     image: Asset.Images.sideMenuActionIconFeedback.image,
-                     discoverabilityTitle: VectorL10n.roomRecentsStartChatWith,
-                     handler: { [weak self] action in
-                self?.startChat()
-            })
-        ])
-        
-        toolbar.items = [
-            UIBarButtonItem(image: UIImage(systemName: "ellipsis.circle"), menu: AllChatsActionProvider().menu),
+        self.tabBarController?.navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(systemName: "line.3.horizontal.decrease.circle"), menu: AllChatsActionProvider().menu)
+    }
+
+    private func updateUI() {
+        let currentSpace = self.dataSource?.currentSpace
+        self.tabBarController?.title = currentSpace?.summary?.displayname ?? VectorL10n.allChatsTitle
+        
+        updateToolbar(with: editActionProvider.updateMenu(with: mainSession, parentSpace: currentSpace, completion: { [weak self] menu in
+            self?.updateToolbar(with: menu)
```

Comment:
I'm confused, this appears to update the toolbar twice with the same menu 🤔

Reply:
This looks odd, I agree, but I can explain 🧐

The menu contains 3 item "Invite people", "Add Room", and "Create subspace" that are available only if the user as the required power level. Fetching the power levels of a room is asynchronous.

My Idea was to implement `AllChatsEditActionProvider.updateMenu()` method so it returns a context menu with all sensible items disabled at first, fetches, in the meantime, the power levels of the space and then calls back the caller with the final menu.

Reply:
Riiight, makes sense now. 👍

Maybe add a couple of comments here and on the `updateMenu` method to that effect?